node/index.ts

```diff
@@ -6,6 +6,18 @@
 import bindings = require('bindings'); // eslint-disable-line @typescript-eslint/no-require-imports
 import * as SignalClient from './libsignal_client';
 
-export const { PrivateKey } = bindings(
-  'libsignal_client'
-) as typeof SignalClient;
+const SC = bindings('libsignal_client') as typeof SignalClient;
+
+const nativeHandle: unique symbol = Symbol();
```

Comment:
tl;dr: this symbol-based approach doesn't achieve true privacy. I'd probably just rely on TypeScript here.

---

I didn't know about `unique symbol`. Cool!

It seems like you're using this symbol as a way to achieve a private class field. This approach is reasonable, but I think there are two better options:

1. Give up on trying to achieve privacy at runtime, and only rely on TypeScript to achieve member privacy. This is what we've done in Signal Desktop and RingRTC, as far as I know. Here's a sketch of how that'd look:

   ```typescript
   export class PrivateKey {
     private readonly nativeHandle: SignalClient.PrivateKey;
     constructor() {
       this.nativeHandle = SC.PrivateKey_generate();
     }
     serialize(): Buffer {
       return SC.PrivateKey_serialize(this.nativeHandle);
     }
   }
   ```

   Of course, this doesn't achieve true privacy, because you could easily do something like this in JavaScript (or TypeScript with enough `any`s and `@ts-ignore`s):

   ```js
   const privateKey = new PrivateKey();
   console.log(privateKey.nativeHandle);
   ```

   Though it's harder, you can still "break into" your symbol-based solution:

   ```js
   const privateKey = new PrivateKey();
   const stolenSymbol = Object.getOwnPropertySymbols(privateKey)[0];
   console.log(privateKey[stolenSymbol]);
   ```
2. Use [JavaScript's private class fields](https://www.valentinog.com/blog/private/) to achieve privacy at runtime and compile time. Here's a sketch:

   ```typescript
   export class PrivateKey {
     readonly #nativeHandle: SignalClient.PrivateKey;
     constructor() {
       this.#nativeHandle = SC.PrivateKey_generate();
     }
     serialize(): Buffer {
       return SC.PrivateKey_serialize(this.#nativeHandle);
     }
   }
   ```

I don't feel strongly about which of those two solutions we use. I'd probably use the latter if I were the tzar of Signal's JavaScript code, but I'd realistically choose the first because we don't use private fields anywhere else (as far as I know).

If I'm mistaken and you're using this symbol for another reason, feel free to ignore this suggestion.

Reply:
I was thinking it made sense to use the same symbol for all of the native-wrapping objects, but I guess they're still opaque on the JavaScript side and not accessible on the Rust side (without the Rust side being the one to access them). And Neon is doing its own type-tagging too. I'll switch to a normal private property.